Issue 77: Add dynamic views

yt-ms opened the pull request:
* [X] fix https://github.com/Midnighter/structurizr-python/issues/77
* [X] description of feature/fix
* [X] tests added/passed
* [X] add an entry to the [next release](../CHANGELOG.rst)

Add support for dynamic views, which show how element collaborate by sequencing interactions.  Based on the [Java implementation](https://github.com/structurizr/java/blob/master/structurizr-core/src/com/structurizr/view/DynamicView.java).

As well as the "parallel" sequence provided in the Java implementation, I've also added a "subsequence" which allows you to get the "1.2", "1.3" style numbering for sub-calls but doesn't seem to be in the Java implementation.

-------------------

THIS SOFTWARE IS CONTRIBUTED SUBJECT TO THE TERMS OF THE Apache License v.2.0. YOU MAY OBTAIN A COPY OF THE LICENSE AT https://www.apache.org/licenses/LICENSE-2.0.

THIS SOFTWARE IS LICENSED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE AND ANY WARRANTY OF NON-INFRINGEMENT, ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. THIS SOFTWARE MAY BE REDISTRIBUTED TO OTHERS ONLY BY EFFECTIVELY USING THIS OR ANOTHER EQUIVALENT DISCLAIMER IN ADDITION TO ANY OTHER REQUIRED LICENSE TERMS.

yt-ms committed:
feat: basic shell for DynamicView

yt-ms committed:
feat: DynamicView hydration

yt-ms committed:
feat: add "response" attribute to RelationshipView to support Dynamic views

yt-ms committed:
feat: add relationships to DynamicView

yt-ms committed:
feat: add auto-incrementing order to DynamicView

yt-ms committed:
feat: uncomment DynamicView in big_bank.py example

yt-ms committed:
feat: add support for parallel sequences in DynamicView

This mirrors the Java implementation, but uses a context manager to
provide cleaner entry/exit to the parallel sequence.

yt-ms committed:
feat: removed the need to specify description when adding relationships to dynamic views

yt-ms committed:
feat: check element is valid to be added to DynamicView

yt-ms committed:
docs: Add to changelog

yt-ms committed:
test: demonstrate issue #79 with xfail

yt-ms committed:
style: fix black and flake8 errors

yt-ms committed:
fix: bug in view.check_parent_and_children_not_in_view()

yt-ms committed:
fix: relationship description shouldn't need to match when adding to DynamicView.

Also patched up big_bank.py example.

yt-ms committed:
test: ensure layout information is copied over for dynamic views

yt-ms committed:
feat: support subsequences in DynamicView

yt-ms committed:
test: missing edge case on DynamicView

yt-ms commented:
I've just spotted that the relationship views need to be ordered when returned, so need to implement that before this PR is merged.

yt-ms committed:
feat: relationship views in DynamicView should be sorted by their order

yt-ms commented:
> I've just spotted that the relationship views need to be ordered when returned, so need to implement that before this PR is merged.

OK, this is done now.

Midnighter commented on examples/big_bank.py:
```diff
@@ -429,14 +429,22 @@ def create_big_bank_workspace():
-    # dynamicView.Add(securityComponent, "select * from users where username = ?",
-    #   database)
-    # dynamicView.PaperSize = PaperSize.A5_Landscape
+    # dynamic diagrams and deployment diagrams are not available with the Free Plan
```
The comment seems obsolete, those features are also part of the free plan now. https://structurizr.com/pricing#comparison

```suggestion
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Provie a Dynamic View."""
```
```suggestion
"""Provide a dynamic view."""
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+
+class DynamicViewIO(ViewIO):
+    """
+    Represent the dynamic view from the C4 model.
```
```suggestion
    Represent a dynamic view on a C4 model.
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+
+class DynamicView(ModelRefMixin, View):
+    """
+    Represent the dynamic view from the C4 model.
```
```suggestion
    Represent a dynamic view on a C4 model.
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+        self,
+        *,
+        software_system: Optional[SoftwareSystem] = None,
+        container: Optional[Container] = None,
```
An alternative is:

```suggestion
        element: Optional[Union[SoftwareSystem, Container]] = None,
```

yt-ms replied:
Done.

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+
+        This will search for a relationship in the model from the source to the
+        destination with matching technology (if specified).  It will also look for
+        situations where this interaction is a "response" in that it it goes in the
```
```suggestion
        situations where this interaction is a "response" in that it goes in the
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+
+        Example of request/response, assuming a single relationship in the model:
+            dynamic_view.add(container1, container2, "Requests data from")
+            dynamic_view.add(container2, container1, "Sends response back to")
```
Let's use the [Google-style docstrings](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) sections.

```suggestion
        Examples:
            Example of a request/response, assuming a single relationship in the model:
            
                dynamic_view.add(container1, container2, "Requests data from")
                dynamic_view.add(container2, container1, "Sends response back to")

```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+                raise ValueError(
+                    f"A relationship between {source.name} and "
+                    f"{destination.name} with technology "
+                    f"'{technology}' does not exist in model."
```
```suggestion
                    f"'{technology}' does not exist in the model."
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+                raise ValueError(
+                    f"A relationship between {source.name} and "
+                    f"{destination.name} does not exist in "
+                    "model."
```
```suggestion
                    "the model."
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+    @contextmanager
+    def subsequence(self):
+        """
+        Start a subsueqnce through a `with` block.
```
```suggestion
        Start a context-managed subsequence.
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+        Start a subsueqnce through a `with` block.
+
+        Subsequences allow nested interaction sequences, showing "child" calls through
+        numbering 1.1, 1.2 etc.  Subsequences can themselves be nested.
```
```suggestion
        numbering 1.1, 1.2, etc.  Subsequences can themselves be nested.
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+
+        Subsequences allow nested interaction sequences, showing "child" calls through
+        numbering 1.1, 1.2 etc.  Subsequences can themselves be nested.
+
```
```suggestion

        Examples:
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+        which in turn calls both C then D to process the request before returning the
+        results back to A.  This can be shown using:
+
+            dynamic_view.add(a, b, "Sends service request to)
```
```suggestion
            dynamic_view.add(a, b, "Sends service request to")
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+            with dynamic_view.subsequence():
+                dynamic_view.add(b, c, "Makes subcall to")
+                dynamic_view.add(b, d, "Makes second subcall to")
+            dynamic_view.add(b, a, "Sends results back to)
```
```suggestion
            dynamic_view.add(b, a, "Sends results back to")
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+    @contextmanager
+    def parallel_sequence(self, continue_numbering: bool):
+        r"""
+        Start a parallel sequence through a `with` block.
```
```suggestion
        Start a context-managed parallel sequence.
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+                                sequence number will continue from after the last
+                                number from the parallel sequence.  If `False` then it
+                                will reset back to the start (usually so you can start
+                                a new parallel sequence).
```
```suggestion
            continue_numbering: Whether to continue the main sequence number
                from where the parallel sequence ended when its context is
                ended (`True`) or to reset the main sequence to where it began
                (`False`). The latter is usually done so that you can start a
                new parallel sequence.
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+                                will reset back to the start (usually so you can start
+                                a new parallel sequence).
+
+        Parallel sequences allow for multiple parallel flows to share the same
```
I'd use the `Examples:` section again. That means indenting everything.

```suggestion
        Examples:
        Parallel sequences allow for multiple parallel flows to share the same
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+
+        Parallel sequences allow for multiple parallel flows to share the same
+        sequence numbers, so e.g.
+                   /-> C -\
```
```suggestion
        sequence numbers, for example,

                   /-> C -\
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+                   /-> C -\
+          A -> B -{        }-> E -> F
+                   \-> D -/
+        could happen concurrently but you want both B->C and B->D to get order
```
```suggestion
                   \-> D -/

        could happen concurrently but you want both B->C and B->D to get order
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+            self.sequence_number.end_subsequence()
+
+    @contextmanager
+    def parallel_sequence(self, continue_numbering: bool):
```
Does it make sense to make one or the other the default value?

Midnighter replied:
@yt-ms a lot of my comments have been hidden.

yt-ms replied:
I think of the two then False is the better default, as you can then make multiple calls and only specify True on the last one (which basically means you're done and can move on).  I'll make that change, and also make it so you have to explicitly name the arg to make it clearer to readers what's going on.

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+
+    @property
+    def relationship_views(self) -> Iterable[RelationshipView]:
+        """Return the relationship views, sorted in order.
```
```suggestion
        """Return the relationship views in order of their sequence number.
```

Midnighter commented on src/structurizr/view/dynamic_view.py:
```diff
@@ -0,0 +1,317 @@
+        description: str,
+        destination: Element,
+        technology: Optional[str],
+    ) -> Tuple[Optional[Relationship], bool]:
```
Missing a docstring. Maybe something like

```
"""Return a pair of the best matching relationship and whether it is a response."""
```

Midnighter commented on src/structurizr/view/view.py:
```diff
@@ -125,6 +125,11 @@ def model(self) -> Model:
+    @property
+    def relationship_views(self) -> Iterable[RelationshipView]:
+        """Return the relationship views contained by this view."""
+        return self._relationship_views
```
Should this return an iterator over the views or a copy of the set?

yt-ms replied:
I've changed to construct a new set - this might confuse folk slightly as in DynamicView it'll be a list, but we're clear here that they should only expect it to be Iterable so I think that's OK.

Midnighter replied:
Sounds good to me.

Midnighter commented on src/structurizr/view/view.py:
```diff
@@ -182,14 +195,21 @@ def _add_relationship(self, relationship: Relationship) -> RelationshipView:
+                    for rv in self._relationship_views
                     if rv.relationship is relationship
+                    and rv.description == description
+                    and rv.response == response
```
Finding a relationship view seems to be a common operation, is it worth adding a method for, maybe?

Midnighter commented on src/structurizr/view/view.py:
```diff
@@ -182,14 +195,21 @@ def _add_relationship(self, relationship: Relationship) -> RelationshipView:
+                )
+                self._relationship_views.add(view)
             return view
         return None
```
This can be removed, `None` is automatically returned.

```suggestion
```

Midnighter commented on src/structurizr/view/view.py:
```diff
@@ -223,7 +247,7 @@ def copy_layout_information_from(self, source: "View") -> None:
                 )
 
-        for source_relationship_view in source.relationship_views:
+        for source_relationship_view in source._relationship_views:
```
"Private" attributes and methods of other objects should not be accessed.

```suggestion
        for source_relationship_view in source.relationship_views:
```

yt-ms replied:
No idea why I changed that!  Reverted.

Midnighter commented on src/structurizr/view/view.py:
```diff
@@ -249,10 +273,23 @@ def find_relationship_view(
         return None
+
+    def check_parent_and_children_not_in_view(self, element: Element) -> None:
+        """Enusre elements can be added with parents or chldren already in the view."""
```
```suggestion
        """Ensure that an element can be added with parents or children already in the view."""
```

yt-ms replied:
Had to tweak the wording to fit within the line length, but should be better now.

Midnighter commented on src/structurizr/view/view.py:
```diff
@@ -249,10 +273,23 @@ def find_relationship_view(
+    def check_parent_and_children_not_in_view(self, element: Element) -> None:
+        """Enusre elements can be added with parents or chldren already in the view."""
+        # Unfortunately not every element has a parent property so we have to scan
+        # The children instead.
```
Your comment makes me think whether a simple test for the parent attribute makes sense?

```py
if hasattr(element, "parent") and view.element is element.parent:
```

or even

```py
parent = getattr(element, "parent", None)
...
if view.element is parent:
```

yt-ms replied:
Yeah, I think that a mix of the two is cleaner than my original:
```python
if view.element is getattr(element, "parent", None):
```

Midnighter commented on src/structurizr/view/view_set.py:
```diff
@@ -321,12 +348,11 @@ def _find_component_view(self, view: ComponentView) -> Optional[ComponentView]:
-    #             return current_view
-    #     return None
+    def _find_dynamic_view(self, view: DynamicView) -> Optional[DynamicView]:
+        for current_view in self.dynamic_views:
+            if view.key == current_view.key:
+                return current_view
+        return None
```
```suggestion
        return next((current_view for current_view in self.dynamic_views if current_view.key == view.key), None)
```

yt-ms replied:
I like it, but when it runs through `black` then it turns into:
```python
        return next(
            (
                current_view
                for current_view in self.dynamic_views
                if current_view.key == view.key
            ),
            None,
        )
```

Is that still better than the original?

Midnighter replied:
Hmm, let's stick with the outer `for` loop but please remove the `return None`. It's superfluous.

yt-ms replied:
Want me to remove the `return None` from the other similar methods whilst I'm here?

Midnighter replied:
Yes, please. I did comment where I saw it but probably missed some 🙂

Midnighter commented on tests/unit/view/test_dynamic_view.py:
```diff
@@ -0,0 +1,420 @@
+
+
+def test_relationship_view_sorting():
+    """Check various cases for sorting relationship views into order."""
```
Would love to see some more cases that test sorting, for example, '1.1' < '1.10' < '1.11' < '10.1' < '10.10'

yt-ms replied:
Done now in `test_view_order.py`.

Midnighter reviewed (changes requested):
Thank you for the hard work on this PR. I did not actually run it but it looks well implemented from reading the code.

I am wondering about one potential design change. What if the `order` attribute's value was not a string but an actual instance of some `SequenceNumber` class? (Not necessarily like the one you have implemented at the moment.) One major benefit that I see is that the ordering is then encapsulated in that class simply the dunder methods `__eq__` and `__lt__`. What do you think?

yt-ms committed:
refactor: introduce class to handle view ordering semantics

yt-ms committed:
refactor: PR feedback

yt-ms commented:
For the ordering of views, I've added a new `ViewOrder` class, which I think helps separate concerns better.  See what you think.

yt-ms committed:
refactor: PR feedback on DynamicView

Rather than specifying software_system= or container= as constructor args, just have element= instead.

yt-ms committed:
refactor: SequenceNumber now returns ViewOrders rather than simple strings

yt-ms committed:
refactor: rename ViewOrder to InteractionOrder

yt-ms committed:
fix: attempt to fix build errors

yt-ms committed:
fix: attempt to fix docs build errors

yt-ms commented on docs/requirements.txt:
```diff
@@ -1,3 +1,4 @@
+livereload ~=2.6.3  # Needed by mkdocstrings
```
Not sure whether this is the correct approach - I would have thought mkdocstrings should bring this in automatically if it needs it.

Midnighter replied:
This is tracked in https://github.com/mkdocstrings/mkdocstrings/issues/295, let's keep an eye on it. Please change the version pin to

```suggestion
livereload ~=2.6
```

yt-ms commented on mkdocs.yml:
```diff
@@ -1,6 +1,7 @@
 site_name: Structurizr-Python
 site_description: A Python client package for the Structurizr cloud service and on-premises installation.
 site_author: Moritz E. Beber
+site_url: https://structurizr-python.readthedocs.io/
```
Not sure whether this also needs `en/latest/` on the end.

yt-ms committed:
fix: update livereload version pin

Midnighter committed:
Update src/structurizr/view/dynamic_view.py

Midnighter committed:
Update src/structurizr/view/dynamic_view.py

Midnighter committed:
Update src/structurizr/view/dynamic_view.py

Midnighter committed:
Update src/structurizr/view/dynamic_view.py

Midnighter committed:
Update src/structurizr/view/dynamic_view.py

Midnighter committed:
Update src/structurizr/view/dynamic_view.py

Midnighter committed:
Update src/structurizr/view/dynamic_view.py

Midnighter committed:
Update src/structurizr/view/dynamic_view.py

Midnighter committed:
Update src/structurizr/view/dynamic_view.py

Midnighter committed:
Update src/structurizr/view/dynamic_view.py

Midnighter committed:
Update src/structurizr/view/dynamic_view.py

Midnighter committed:
Apply suggestions from code review

Add changes to docstrings.

yt-ms committed:
fix: PR feedback

yt-ms committed:
fix: PR feedback

yt-ms committed:
refactor: make view.find_element_view and find_relationship_view more generally useful

yt-ms committed:
docs: update CHANGELOG with breaking API change

Midnighter commented:
What a monster work! Thanks a lot 🙂
